lib/APMSpan.js

```diff
@@ -29,43 +32,195 @@ const LOGS = Symbol('LOGS');
+                return undefined;
+            }
+
+            if (childOf.length === 1) {
+                return childOf[0];
+            } else if (childOf.length > 1) {
+                return undefined;
+            }
+
+            if (followsFrom.length === 1) {
+                return followsFrom[0];
+            }
+            return undefined;
+        }
+
+        // standardise reference.referencedContext() to context.
+        if (references) {
+            references = references.map(item =>
+                new Constants.Reference(item.type(), item.referencedContext() instanceof APMSpan ?
+                    item.referencedContext().context() : item.referencedContext())
+            );
+        }
+
+        if (!references || references.length === 0) {
+            this._initChildOf(references);
+            return;
+        }
+
+        const primaryReference = findPrimaryReference(references);
+
+        if (primaryReference) {
+            if (primaryReference.referencedContext().getConsumerCorrelationId()) {
+                primaryReference._type = Constants.REFERENCE_CHILD_OF;
+            }
+
+            switch (primaryReference.type()) {
+            case Constants.REFERENCE_CHILD_OF:
+                this._initChildOf(primaryReference);
+                break;
+            case Constants.REFERENCE_FOLLOWS_FROM:
+                this._initFollowsFrom(primaryReference);
+                break;
+            default:
+                console.log('Undefined causal reference type');
+            }
+        } else {
+            this._initJoin(references);
+        }
     }
 
-    _initSpanContext(parentSpanContext) {
+    _initChildOf(reference) {
+        const spanContext = reference ? reference.referencedContext() : undefined;
+
         const id = utils.generateSpanId();
         let parentId;
         let traceId = utils.generateSpanId();
 
-        if (parentSpanContext && parentSpanContext.getId() && parentSpanContext.getTraceId()) {
-            parentId = parentSpanContext.getId();
-            traceId = parentSpanContext.getTraceId();
+        if (spanContext && spanContext.getId() && spanContext.getTraceId()) {
+            parentId = spanContext.getId();
+            traceId = spanContext.getTraceId();
         } else {
+            // new trace
             traceId = id;
         }
 
-        this[SPAN_CONTEXT] = new APMSpanContext(id, traceId, parentId, parentSpanContext);
-
-        let nodeType = 'Component';
-        const consumerCorrelationId = parentSpanContext && parentSpanContext.getConsumerCorrelationId();
+        let nodeType = NODE_TYPE_COMPONENT;
+        this[SPAN_CONTEXT] = new APMSpanContext(id, traceId, parentId, spanContext);
+        const consumerCorrelationId = spanContext && spanContext.getConsumerCorrelationId();
         if (consumerCorrelationId !== undefined) {
-            nodeType = 'Consumer';
+            nodeType = NODE_TYPE_CONSUMER;
```

Comment:
Ah ok - might be worth adding a test just to be sure. I will add one for the Java provider, as I don't think I check for that.